### PR TITLE
Add naming conventions.

### PR DESCRIPTION
**Pull request by dourouc05**
Follows #1459.

### CODE REVIEW

File: docs/src/background/naming_conventions.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+## Sets
+
+Sets encode the structure of constraints. Their names should follow the 
+following conventions: 
+
+* The name of concrete sets does not end with `Set`, but the name of abstract
+  sets end with `Set`. For instance, `MOI.ZeroOne` is a concrete set
+  (it can be instanciated by the user), but `MOI.AbstractScalarSet` is an 
+  abstract set (many sets should have it as supertype; an 
+  `MOI.AbstractScalarSet` cannot be instanciated).
```
Comment (odow) [inline docs/src/background/naming_conventions.md:15]:
```suggestion
  `MOI.AbstractScalarSet` cannot be instantiated).
```

File: docs/src/background/naming_conventions.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  the matrix is symmetric, whose name ends with `ConeTriangle` 
+  (like `MOI.PositiveSemidefiniteConeTriangle`).
+* The name of the set is singular, instead of plural: prefer `Integer` to 
+  `Integers`.
+* The sets are named with nouns instead of verbs. Usual sets in mixed-integer
+  and convex programming only have such names.
+* As much as possible, the names should follow established conventions in the 
+  domain where this set is used: for instance, convex sets should have names 
+  close to those of CVX, and constraint-programming sets should follow 
+  MiniZinc's constraints.
```
Comment (odow) [inline docs/src/background/naming_conventions.md:30]: Maybe include a link to their documentation?

**Comment by dourouc05** [comment]
Thanks for having a look :)!

**Comment by mlubin** [comment]
It looks like the discussion is in #1459, so I responded there.

**Review by odow — COMMENTED** [review]
@blegat suggests Cartesian products are plural, without "cone"

File: docs/src/background/naming_conventions.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
+# Naming conventions
+
+MOI follows several conventions for naming functions and structures. Ideally,
```
Comment (odow) [inline docs/src/background/naming_conventions.md:3]:
```suggestion
MOI follows several conventions for naming functions and structures.
```
No need for "Ideally,"

File: docs/src/background/naming_conventions.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+## Sets
+
+Sets encode the structure of constraints. Their names should follow the 
+following conventions: 
+
+* The name of concrete sets does not end with `Set`, but the name of abstract
+  sets end with `Set`. For instance, `MOI.ZeroOne` is a concrete set
+  (it can be instantiated by the user), but `MOI.AbstractScalarSet` is an 
+  abstract set (many sets should have it as supertype; an 
+  `MOI.AbstractScalarSet` cannot be instantiated).
```
Comment (odow) [inline docs/src/background/naming_conventions.md:15]: Prefer to be more direct.
```
 * Abstract types in the set hierarchy should begin with `Abstract` and end in
   `Set`, e.g., [`MOI.AbstractScalarSet`](@ref), [`MOI.AbstractVectorSet`](@ref).
```

File: docs/src/background/naming_conventions.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+* The name of conic sets ends with `Cone`, like `NormInfinityCone`, unless 
+  the set is matrix-valued.
```
Comment (odow) [inline docs/src/background/naming_conventions.md:17]:
```
 * Vector-valued conic sets should end with `Cone`, e.g.,
   [`MOI.NormInfinityCone`](@ref), [`MOI.SecondOrderCone`](@ref).
 * Vector-valued cartesian products should be plural and not end in `Cone`,
   e.g., [`MOI.Nonnegatives`](@ref) not `MOI.NonnegativeCone`.
```

File: docs/src/background/naming_conventions.md [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  (it can be instantiated by the user), but `MOI.AbstractScalarSet` is an 
+  abstract set (many sets should have it as supertype; an 
+  `MOI.AbstractScalarSet` cannot be instantiated).
+* The name of conic sets ends with `Cone`, like `NormInfinityCone`, unless 
+  the set is matrix-valued. 
+* Matrix-valued conic sets provide two versions: one for a standard 
+  rectangular representation, whose name ends with `ConeSquare` (like
+  `MOI.LogDetConeSquare`); one with a packed representation considering that
+  the matrix is symmetric, whose name ends with `ConeTriangle` 
+  (like `MOI.PositiveSemidefiniteConeTriangle`).
```
Comment (odow) [inline docs/src/background/naming_conventions.md:22]:
```
 * Matrix-valued conic sets should provide two representations: `ConeSquare` and
   `ConeTriangle`, e.g., [`MOI.RootDetConeTriangle`](@ref) and
   [`MOI.RootDetConeSquare`](@ref). See [Matrix cones](@ref) for more details.
```

File: docs/src/background/naming_conventions.md [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+* The name of the set is singular, instead of plural: prefer `Integer` to 
+  `Integers`.
```
Comment (odow) [inline docs/src/background/naming_conventions.md:24]:
```
 * Scalar sets should be singular, not plural, e.g., [`MOI.Integer`](@ref), not 
   `MOI.Integers`.
```

File: docs/src/background/naming_conventions.md [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+* The sets are named with nouns instead of verbs. Usual sets in mixed-integer
+  and convex programming only have such names.
```
Comment (odow) [inline docs/src/background/naming_conventions.md:26]: What about `Complements`? I think we should just remove this one.

Comment (dourouc05) [reply docs/src/background/naming_conventions.md:23]: Having a similar rule would really help with CP sets. Plus, I'm not sure that `Complements` is a good name in this case (https://github.com/jump-dev/MathOptInterface.jl/issues/1459#issuecomment-882091845): typically, `complements` is used as an infix operator in modelling languages, but MOI doesn't have "infix sets".

File: docs/src/background/naming_conventions.md [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+MOI follows several conventions for naming functions and structures. These 
+should also be followed by packages extending MOI.
+
+## Sets
+
+Sets encode the structure of constraints. Their names should follow the 
+following conventions: 
+
+* Abstract types in the set hierarchy should begin with `Abstract` and end in
+  `Set`, e.g., [`MOI.AbstractScalarSet`](@ref), [`MOI.AbstractVectorSet`](@ref).
```
Comment (odow) [inline docs/src/background/naming_conventions.md:12]:
```suggestion
  `Set`, e.g., [`AbstractScalarSet`](@ref), [`AbstractVectorSet`](@ref).
```

File: docs/src/background/naming_conventions.md [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+* Vector-valued conic sets should end with `Cone`, e.g.,
+  [`MOI.NormInfinityCone`](@ref), [`MOI.SecondOrderCone`](@ref).
```
Comment (odow) [inline docs/src/background/naming_conventions.md:14]:
```suggestion
  [`NormInfinityCone`](@ref), [`SecondOrderCone`](@ref).
```

File: docs/src/background/naming_conventions.md [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+* Vector-valued Cartesian products should be plural and not end in `Cone`,
+  e.g., [`MOI.Nonnegatives`](@ref), not `MOI.NonnegativeCone`.
```
Comment (odow) [inline docs/src/background/naming_conventions.md:16]:
```suggestion
  e.g., [`Nonnegatives`](@ref), not `NonnegativeCone`.
```

File: docs/src/background/naming_conventions.md [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+* Matrix-valued conic sets should provide two representations: `ConeSquare` and
+  `ConeTriangle`, e.g., [`MOI.RootDetConeTriangle`](@ref) and
```
Comment (odow) [inline docs/src/background/naming_conventions.md:18]:
```suggestion
  `ConeTriangle`, e.g., [`RootDetConeTriangle`](@ref) and
```

File: docs/src/background/naming_conventions.md [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+  [`MOI.RootDetConeSquare`](@ref). See [Matrix cones](@ref) for more details.
```
Comment (odow) [inline docs/src/background/naming_conventions.md:19]:
```suggestion
  [`RootDetConeSquare`](@ref). See [Matrix cones](@ref) for more details.
```

File: docs/src/background/naming_conventions.md [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+* Scalar sets should be singular, not plural, e.g., [`MOI.Integer`](@ref), not
```
Comment (odow) [inline docs/src/background/naming_conventions.md:20]:
```suggestion
* Scalar sets should be singular, not plural, e.g., [`Integer`](@ref), not 
```

File: docs/src/background/naming_conventions.md [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+  `Set`, e.g., [`MOI.AbstractScalarSet`](@ref), [`MOI.AbstractVectorSet`](@ref).
+* Vector-valued conic sets should end with `Cone`, e.g.,
+  [`MOI.NormInfinityCone`](@ref), [`MOI.SecondOrderCone`](@ref).
+* Vector-valued Cartesian products should be plural and not end in `Cone`,
+  e.g., [`MOI.Nonnegatives`](@ref), not `MOI.NonnegativeCone`.
+* Matrix-valued conic sets should provide two representations: `ConeSquare` and
+  `ConeTriangle`, e.g., [`MOI.RootDetConeTriangle`](@ref) and
+  [`MOI.RootDetConeSquare`](@ref). See [Matrix cones](@ref) for more details.
+* Scalar sets should be singular, not plural, e.g., [`MOI.Integer`](@ref), not 
+  `MOI.Integers`.
```
Comment (odow) [inline docs/src/background/naming_conventions.md:21]:
```suggestion
  `Integers`.
```

**Review by blegat — APPROVED** [review]

**Comment by odow** [comment]
Merging these for now. We can always refine later

**Pull request merged**